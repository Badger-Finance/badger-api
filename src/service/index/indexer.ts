import {
	EventInput,
	getBlock,
	getIndexedBlock,
	GetPriceFunc,
	getSett,
	respond,
	saveItem,
	THIRTY_MIN_BLOCKS,
} from '../util/util';

const ASSET_DATA = process.env.ASSET_DATA || ''; // FIXME: sane defaults?

export const indexAsset = async (event: EventInput, getPrice: GetPriceFunc) => {
	const { asset, createdBlock, contract } = event;
	let block = await getIndexedBlock(ASSET_DATA, asset, createdBlock);
	console.log(`Index ${asset} at height: ${block}`);

	while (true) {
		const sett = await getSett(contract, block);

		if (sett.errors != undefined) {
			break;
		}

		if (sett.data == null || sett.data.sett == null) {
			block += THIRTY_MIN_BLOCKS;
			continue;
		}

		const { balance, totalSupply, pricePerFullShare } = sett.data.sett;

		const decimals = asset === 'digg' ? 9 : 18;
		const blockData = await getBlock(block);
		const timestamp = Number(blockData.timestamp) * 1000;
		const balanceComputed = balance / Math.pow(10, decimals);
		const supply = totalSupply / Math.pow(10, decimals);
		const ratio = pricePerFullShare / Math.pow(10, 18);
		const value = balanceComputed * (await getPrice(sett));

		const snapshot = {
			asset: asset,
			height: block,
			timestamp: timestamp,
			balance: balance,
			supply: supply,
			ratio: ratio,
			value: parseFloat(value.toFixed(2)),
		};

		await saveItem(ASSET_DATA, snapshot);
		block += THIRTY_MIN_BLOCKS;
	}

	return respond(200); // FIXME: is this right?
};
