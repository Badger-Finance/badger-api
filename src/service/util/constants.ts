export const TOKENS = {
	// general tokens
	RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
	BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
	TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
	SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',
	DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',

	// uniswap tokens
	UNI_BADGER: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
	UNI_DIGG: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

	// sushiswap tokens
	SUSHI_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
	SUSHI_BADGER: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
	SUSHI_DIGG: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',
};
