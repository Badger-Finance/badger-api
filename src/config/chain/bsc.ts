import { SettDefinition } from '../../interface/Sett';
import { Protocol, TOKENS } from '../constants';

export const bscSetts: SettDefinition[] = [
  {
    name: 'Pancakeswap BNB/BTCB',
    symbol: 'PLP-BNB-BTCB',
    depositToken: TOKENS.PANCAKE_BNB_BTCB,
    settToken: '0xF6BC36280F32398A031A7294e81131aEE787D178',
    protocol: Protocol.Pancakeswap,
  },
  {
    name: 'Pancakeswap bBADGER/BTCB',
    symbol: 'PLP-BBADGER-BTCB',
    depositToken: TOKENS.PANCAKE_BBADGER_BTCB,
    settToken: '0xba89fec5081ce1cdf24ab897fb80f6e9e6d3fda1',
    protocol: Protocol.Pancakeswap,
  },
  {
    name: 'Pancakeswap bDIGG/BTCB',
    symbol: 'PLP-BDIGG-BTCB',
    depositToken: TOKENS.PANCAKE_BDIGG_BTCB,
    settToken: '0xa71ebba5f3f24e84be96240264ae5de38b63860d',
    protocol: Protocol.Pancakeswap,
  },
];
