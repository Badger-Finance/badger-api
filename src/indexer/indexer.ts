import { ASSET_DATA } from '../util/constants';
import {
	EventInput,
	getBlock,
	getIndexedBlock,
	GetPriceFunc,
	getSett,
	saveItem,
	THIRTY_MIN_BLOCKS,
} from '../util/util';

export const indexAsset = async (event: EventInput, getPrice: GetPriceFunc) => {
	const { asset, createdBlock, contract } = event;
	let block = await getIndexedBlock(ASSET_DATA, asset, createdBlock);
	console.log(`Index ${asset} at height: ${block}`);

	while (true) {
		const sett = await getSett(contract, block);

		if (sett.errors != undefined) {
			break;
		}

		if (sett.data == null || sett.data.sett == null) {
			block += THIRTY_MIN_BLOCKS;
			continue;
		}

		const { balance, totalSupply, pricePerFullShare, token } = sett.data.sett;

		const blockData = await getBlock(block);
		const timestamp = Number(blockData.timestamp) * 1000;
		const tokenBalance = balance / Math.pow(10, token.decimals);
		const supply = totalSupply / Math.pow(10, token.decimals);
		const ratio = pricePerFullShare / Math.pow(10, 18);
		const value = tokenBalance * (await getPrice(sett)).usd;

		const snapshot = {
			asset: asset,
			height: block,
			timestamp: timestamp,
			balance: balance,
			supply: supply,
			ratio: ratio,
			value: parseFloat(value.toFixed(2)),
		};

		await saveItem(ASSET_DATA, snapshot);
		block += THIRTY_MIN_BLOCKS;
	}
};
