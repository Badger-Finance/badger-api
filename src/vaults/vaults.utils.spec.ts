import BadgerSDK, {
  BadgerGraph,
  Protocol,
  TokensService,
  VaultBehavior,
  VaultDTO,
  VaultsService,
  VaultState,
  VaultType,
} from '@badger-dao/sdk';
import { BadRequest, NotFound } from '@tsed/exceptions';
import { BigNumber, ethers } from 'ethers';
import { BinanceSmartChain } from '../chains/config/bsc.config';
import { TOKENS } from '../config/tokens.config';
import { BouncerType } from '../rewards/enums/bouncer-type.enum';
import {
  randomVault,
  randomSnapshot,
  setupMapper,
  TEST_CHAIN,
  TEST_ADDR,
  randomSnapshots,
  setFullTokenDataMock,
} from '../test/tests.utils';
import {
  defaultVault,
  estimateDerivativeEmission,
  getCachedVault,
  getVaultDefinition,
  getVaultPerformance,
  getVaultTokenPrice,
  getVaultUnderlyingPerformance,
  VAULT_SOURCE,
} from './vaults.utils';
import * as pricesUtils from '../prices/prices.utils';
import * as rewardsUtils from '../rewards/rewards.utils';
import * as protocolsUtils from '../protocols/protocols.utils';
import * as indexerUtils from '../indexers/indexer.utils';
import { createValueSource } from '../protocols/interfaces/value-source.interface';
import { tokenEmission } from '../protocols/protocols.utils';
import { Polygon } from '../chains/config/polygon.config';
import { SourceType } from '../rewards/enums/source-type.enum';
import { ONE_DAY_SECONDS, ONE_YEAR_MS } from '../config/constants';
import { fullTokenMockMap } from '../tokens/mocks/full-token.mock';
import { TokenNotFound } from '../tokens/errors/token.error';

describe('vaults.utils', () => {
  const vault = getVaultDefinition(TEST_CHAIN, TOKENS.BBADGER);

  beforeEach(() => {
    jest.spyOn(BadgerGraph.prototype, 'loadSettHarvests').mockImplementation(async (_options) => {
      const harvests = [
        {
          id: '',
          blockNumber: '10',
          timestamp: 1640159886,
          amount: '80000000000000000000',
          token: {
            id: '',
            symbol: '',
            name: '',
            decimals: '',
            totalSupply: '',
          },
        },
        {
          id: '',
          blockNumber: '10',
          timestamp: 1642159886,
          amount: '100000000000000000000',
          token: {
            id: '',
            symbol: '',
            name: '',
            decimals: '',
            totalSupply: '',
          },
        },
        {
          id: '',
          blockNumber: '10',
          timestamp: 1645159886,
          amount: '125000000000000000000',
          token: {
            id: '',
            symbol: '',
            name: '',
            decimals: '',
            totalSupply: '',
          },
        },
      ];
      return {
        settHarvests: harvests,
      };
    });
    jest.spyOn(BadgerGraph.prototype, 'loadBadgerTreeDistributions').mockImplementation(async () => ({
      badgerTreeDistributions: [],
    }));
    jest.spyOn(TokensService.prototype, 'loadToken').mockImplementation(async (address) => ({
      address,
      decimals: 18,
      symbol: 'TEST',
      name: 'TEST',
    }));
    const snapshot = randomSnapshot(vault);
    snapshot.value = 1000;
    snapshot.balance = 10000;
    setupMapper([snapshot]);
    jest.spyOn(indexerUtils, 'getVault').mockImplementation(async (_chain, _address) => ({
      sett: {
        id: TEST_ADDR,
        balance: '2500000000000000000000', // 2500
        available: 0,
        netDeposit: 0,
        netShareDeposit: 0,
        token: {
          symbol: 'TEST',
          name: 'TEST',
          id: TEST_ADDR,
          decimals: 18,
          totalSupply: 3,
        },
        pricePerFullShare: 1,
        totalSupply: 10,
        symbol: 'TEST',
        name: 'TEST',
        decimals: 18,
        grossDeposit: 1,
        grossShareDeposit: 1,
        grossShareWithdraw: 1,
        grossWithdraw: 1,
        strategy: {
          id: TEST_ADDR,
          balance: '2500000000000000000000',
        },
      },
    }));
    jest.spyOn(rewardsUtils, 'getProtocolValueSources').mockImplementation(async (_chain, _vault) => {
      const rewardSource = createValueSource('Test LP Fees', 1.13);
      return [rewardsUtils.valueSourceToCachedValueSource(rewardSource, vault, SourceType.TradeFee)];
    });
  });

  function setupSdk() {
    jest.spyOn(VaultsService.prototype, 'listHarvests').mockImplementation(async (opts) => {
      if (!opts.timestamp_gte) {
        throw new Error('Invalid request!');
      }
      const startTime = opts.timestamp_gte;
      const data = [0, 1, 2].map((int) => {
        const timestamp = Number((startTime + int * ONE_DAY_SECONDS * 14).toFixed());
        const block = Number((timestamp / 10000).toFixed());
        return {
          timestamp,
          harvests: [
            { timestamp, block, token: TOKENS.CRV_IBBTC, amount: BigNumber.from((int + 1 * 1.88e18).toString()) },
          ],
          treeDistributions: [
            { timestamp, block, token: TOKENS.BCVXCRV, amount: BigNumber.from((int + 1 * 5.77e12).toString()) },
            { timestamp, block, token: TOKENS.BVECVX, amount: BigNumber.from((int + 1 * 4.42e12).toString()) },
          ],
        };
      });
      return { data };
    });
  }

  beforeEach(() => {
    jest.spyOn(BadgerSDK.prototype, 'ready').mockImplementation();
    const vault = getVaultDefinition(TEST_CHAIN, TOKENS.BBADGER);
    jest.spyOn(rewardsUtils, 'getRewardEmission').mockImplementation(async (_chain, _vault) => {
      const rewardSource = createValueSource('Badger Rewards', 6.969);
      return [
        rewardsUtils.valueSourceToCachedValueSource(
          rewardSource,
          vault,
          tokenEmission(fullTokenMockMap[TOKENS.BBADGER], true),
        ),
      ];
    });
  });

  describe('defaultVault', () => {
    it('returns a sett default fields', async () => {
      const vaultDefinition = randomVault();

      const depositToken = fullTokenMockMap[vaultDefinition.depositToken];
      const settToken = fullTokenMockMap[vaultDefinition.vaultToken];
      const expected: VaultDTO = {
        asset: depositToken.symbol,
        vaultAsset: settToken.symbol,
        state: vaultDefinition.state
          ? vaultDefinition.state
          : vaultDefinition.newVault
          ? VaultState.New
          : VaultState.Open,
        apr: 0,
        apy: 0,
        balance: 0,
        available: 0,
        boost: {
          enabled: false,
          weight: 0,
        },
        bouncer: vaultDefinition.bouncer ?? BouncerType.None,
        name: vaultDefinition.name,
        protocol: Protocol.Badger,
        pricePerFullShare: 1,
        sources: [],
        sourcesApy: [],
        tokens: [],
        underlyingToken: vaultDefinition.depositToken,
        value: 0,
        vaultToken: vaultDefinition.vaultToken,
        strategy: {
          address: ethers.constants.AddressZero,
          withdrawFee: 50,
          performanceFee: 20,
          strategistFee: 0,
        },
        type: vaultDefinition.protocol === Protocol.Badger ? VaultType.Native : VaultType.Standard,
        behavior: vaultDefinition.behavior ?? VaultBehavior.None,
        lastHarvest: 0,
        yieldProjection: {
          yieldApr: 0,
          yieldTokens: [],
          yieldValue: 0,
          harvestApr: 0,
          harvestApy: 0,
          harvestTokens: [],
          harvestValue: 0,
        },
      };
      setFullTokenDataMock();
      const actual = await defaultVault(TEST_CHAIN, vaultDefinition);
      expect(actual).toMatchObject(expected);
    });
  });

  describe('getCachedVault', () => {
    describe('no cached vault exists', () => {
      it('returns the default sett', async () => {
        setupMapper([]);
        const vaultDefinition = randomVault();
        setFullTokenDataMock();
        const cached = await getCachedVault(TEST_CHAIN, vaultDefinition);
        const defaultVaultInst = await defaultVault(TEST_CHAIN, vaultDefinition);
        expect(cached).toMatchObject(defaultVaultInst);
      });
    });

    describe('a cached vault exists', () => {
      it('returns the vault', async () => {
        const vault = randomVault();
        const snapshot = randomSnapshot(vault);
        setupMapper([snapshot]);
        setFullTokenDataMock();

        const cached = await getCachedVault(TEST_CHAIN, vault);
        const expected = await defaultVault(TEST_CHAIN, vault);
        expected.pricePerFullShare = snapshot.balance / snapshot.totalSupply;
        expected.balance = snapshot.balance;
        expected.value = snapshot.value;
        expected.boost = {
          enabled: snapshot.boostWeight > 0,
          weight: snapshot.boostWeight,
        };
        expect(cached).toMatchObject(expected);
      });
    });
  });

  describe('getVaultDefinition', () => {
    describe('for an existing sett', () => {
      it('returns the expected sett definition', () => {
        const expected = TEST_CHAIN.vaults[Math.floor(Math.random() * TEST_CHAIN.vaults.length)];
        const actual = getVaultDefinition(TEST_CHAIN, expected.vaultToken);
        expect(actual).toMatchObject(expected);
      });
    });

    describe('for an existing sett on a different chain', () => {
      it('throws a not found error', () => {
        const expected = TEST_CHAIN.vaults[Math.floor(Math.random() * TEST_CHAIN.vaults.length)];
        expect(() => getVaultDefinition(new BinanceSmartChain(), expected.vaultToken)).toThrow(NotFound);
      });
    });

    describe('for an non existing sett', () => {
      it('throws a not found error', () => {
        const expected = TEST_CHAIN.vaults[Math.floor(Math.random() * TEST_CHAIN.vaults.length)];
        expect(() => getVaultDefinition(TEST_CHAIN, expected.depositToken)).toThrow(NotFound);
      });
    });
  });

  describe('getVaultTokenPrice', () => {
    describe('look up non vault token price', () => {
      it('throws a bad request error', async () => {
        setFullTokenDataMock();
        await expect(getVaultTokenPrice(TEST_CHAIN, TOKENS.BADGER)).rejects.toThrow(BadRequest);
      });
    });

    describe('look up malformed token configuration', () => {
      it('throws an unprocessable entity error', async () => {
        setFullTokenDataMock();
        await expect(getVaultTokenPrice(TEST_CHAIN, TEST_ADDR)).rejects.toThrow(TokenNotFound);
      });
    });

    describe('look up valid, properly configured vault', () => {
      it('returns a valid token price for the vault base on price per full share', async () => {
        const vault = randomVault(TEST_CHAIN);
        const snapshot = randomSnapshot(vault);
        setupMapper([snapshot]);
        jest.spyOn(pricesUtils, 'getPrice').mockImplementation(async (address) => ({ address, price: 10 }));
        setFullTokenDataMock();
        const vaultPrice = await getVaultTokenPrice(TEST_CHAIN, vault.vaultToken);
        expect(vaultPrice).toMatchObject({
          address: vault.vaultToken,
          price: 10 * snapshot.pricePerFullShare,
        });
      });
    });
  });

  describe('getVaultPerformance', () => {
    const vault = getVaultDefinition(TEST_CHAIN, TOKENS.BBADGER);

    describe('no rewards or harvests', () => {
      it('returns value sources from fallback methods', async () => {
        jest.spyOn(VaultsService.prototype, 'listHarvests').mockImplementation(async (_opts) => ({ data: [] }));
        setFullTokenDataMock();
        const result = await getVaultPerformance(TEST_CHAIN, vault);
        expect(result).toMatchSnapshot();
      });
    });

    describe('requests non standard vault performance', () => {
      it('returns value sources from fallback methods', async () => {
        jest.spyOn(VaultsService.prototype, 'listHarvests').mockImplementation(async (_opts) => {
          throw new Error('Incompatible vault!');
        });
        setupMapper(randomSnapshots(vault));
        setFullTokenDataMock();
        const result = await getVaultPerformance(TEST_CHAIN, vault);
        expect(result).toMatchSnapshot();
      });
    });

    describe('requests non compatible network vault performances', () => {
      it('returns value sources from fallback methods', async () => {
        const alternateChain = new Polygon();
        const vault = getVaultDefinition(alternateChain, TOKENS.BMATIC_QUICK_USDC_WBTC);
        setFullTokenDataMock();
        const result = await getVaultPerformance(alternateChain, vault);
        expect(result).toMatchSnapshot();
      });
    });

    describe('requests standard vault performance', () => {
      it('returns value sources from standard methods', async () => {
        setupSdk();
        jest.spyOn(pricesUtils, 'getPrice').mockImplementation(async (token) => ({
          address: token,
          price: Number(token.slice(0, 4)),
        }));
        jest.spyOn(protocolsUtils, 'getVaultCachedValueSources').mockImplementation(async (vault) => {
          const underlying = createValueSource(VAULT_SOURCE, 10);
          return [rewardsUtils.valueSourceToCachedValueSource(underlying, vault, SourceType.PreCompound)];
        });
        setFullTokenDataMock();
        const result = await getVaultPerformance(TEST_CHAIN, vault);
        expect(result).toMatchSnapshot();
      });

      it('skips all emitted tokens with no price', async () => {
        setupSdk();
        jest.spyOn(pricesUtils, 'getPrice').mockImplementation(async (token) => {
          if (token !== vault.depositToken) {
            return {
              address: token,
              price: 0,
            };
          }
          return {
            address: token,
            price: Number(token.slice(0, 4)),
          };
        });
        setFullTokenDataMock();
        const result = await getVaultPerformance(TEST_CHAIN, vault);
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('estimateDerivativeEmission', () => {
    it.each([
      // enumerate all test cases above / below 100% apr
      [3.4883, 6.7204, 6.7204, 94.20314377878076],
      [3.4883, 6.7204, 0.7204, 12.45153507752425],
      [3.4883, 0.7204, 6.7204, 80.69651967371782],
      [3.4883, 0.7204, 0.7204, 3.529515865690866],
      [0.4883, 6.7204, 6.7204, 98.7771522235121],
      [0.4883, 6.7204, 0.7204, 26.1556767065858],
      [0.4883, 0.7204, 0.7204, 13.257249960438303],
    ])(
      'Estimates derived emission from (%d compound, %d emission, %d compound emission) as %d%%',
      (compound, emission, compoundEmission, expected) => {
        expect(estimateDerivativeEmission(compound, emission, compoundEmission)).toEqual(expected);
      },
    );
  });

  describe('getVaultUnderlyingPerformance', () => {
    it('returns 0 for no pricePerFullShare increase', async () => {
      const vault = randomVault();
      const snapshot = randomSnapshot(vault);
      setupMapper([snapshot]);
      setFullTokenDataMock();
      const result = await getVaultUnderlyingPerformance(TEST_CHAIN, vault);
      result.forEach((r) => expect(r.apr).toEqual(0));
    });

    it('returns expected apr for increase in pricePerFullShare', async () => {
      const vault = randomVault();
      const snapshots = randomSnapshots(vault);
      setupMapper(snapshots);
      const duration = snapshots[0].timestamp - snapshots[snapshots.length - 1].timestamp;
      const deltaPpfs = snapshots[0].pricePerFullShare - snapshots[snapshots.length - 1].pricePerFullShare;
      const expected = (deltaPpfs / snapshots[snapshots.length - 1].pricePerFullShare) * (ONE_YEAR_MS / duration) * 100;
      setFullTokenDataMock();
      const result = await getVaultUnderlyingPerformance(TEST_CHAIN, vault);
      result.forEach((r) => expect(r.apr).toEqual(expected));
    });
  });
});
