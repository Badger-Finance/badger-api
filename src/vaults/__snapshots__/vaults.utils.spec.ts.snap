// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vaults.utils getVaultPerformance no rewards or harvests returns value sources from fallback methods 1`] = `
Array [
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_compound",
    "apr": 189.11917098445593,
    "boostable": false,
    "harvestable": false,
    "maxApr": 189.11917098445593,
    "minApr": 189.11917098445593,
    "name": "Vault Compounding",
    "oneDay": 183.4170854271372,
    "sevenDay": 189.11917098445593,
    "thirtyDay": 214.7058823529413,
    "threeDay": 185.27918781725862,
    "type": "compound",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_flat_Sushi_emission",
    "apr": 8.888,
    "boostable": false,
    "harvestable": false,
    "maxApr": 8.888,
    "minApr": 8.888,
    "name": "Sushi Rewards",
    "oneDay": 8.888,
    "sevenDay": 8.888,
    "thirtyDay": 8.888,
    "threeDay": 8.888,
    "type": "flat_Sushi_emission",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_boosted_bBADGER_emission",
    "apr": 6.969,
    "boostable": false,
    "harvestable": false,
    "maxApr": 6.969,
    "minApr": 6.969,
    "name": "Badger Rewards",
    "oneDay": 6.969,
    "sevenDay": 6.969,
    "thirtyDay": 6.969,
    "threeDay": 6.969,
    "type": "boosted_bBADGER_emission",
  },
]
`;

exports[`vaults.utils getVaultPerformance requests non compatible network vault performances returns value sources from fallback methods 1`] = `
Array [
  CachedValueSource {
    "address": "0x6B2d4c4bb50274c5D4986Ff678cC971c0260E967",
    "addressValueSourceType": "0x6B2d4c4bb50274c5D4986Ff678cC971c0260E967_compound",
    "apr": 189.11917098445593,
    "boostable": false,
    "harvestable": false,
    "maxApr": 189.11917098445593,
    "minApr": 189.11917098445593,
    "name": "Vault Compounding",
    "oneDay": 183.4170854271372,
    "sevenDay": 189.11917098445593,
    "thirtyDay": 214.7058823529413,
    "threeDay": 185.27918781725862,
    "type": "compound",
  },
  CachedValueSource {
    "address": "0x6B2d4c4bb50274c5D4986Ff678cC971c0260E967",
    "addressValueSourceType": "0x6B2d4c4bb50274c5D4986Ff678cC971c0260E967_flat_Sushi_emission",
    "apr": 8.888,
    "boostable": false,
    "harvestable": false,
    "maxApr": 8.888,
    "minApr": 8.888,
    "name": "Sushi Rewards",
    "oneDay": 8.888,
    "sevenDay": 8.888,
    "thirtyDay": 8.888,
    "threeDay": 8.888,
    "type": "flat_Sushi_emission",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_boosted_bBADGER_emission",
    "apr": 6.969,
    "boostable": false,
    "harvestable": false,
    "maxApr": 6.969,
    "minApr": 6.969,
    "name": "Badger Rewards",
    "oneDay": 6.969,
    "sevenDay": 6.969,
    "thirtyDay": 6.969,
    "threeDay": 6.969,
    "type": "boosted_bBADGER_emission",
  },
]
`;

exports[`vaults.utils getVaultPerformance requests non standard vault performance returns value sources from fallback methods 1`] = `
Array [
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_compound",
    "apr": 189.11917098445593,
    "boostable": false,
    "harvestable": false,
    "maxApr": 189.11917098445593,
    "minApr": 189.11917098445593,
    "name": "Vault Compounding",
    "oneDay": 183.4170854271372,
    "sevenDay": 189.11917098445593,
    "thirtyDay": 214.7058823529413,
    "threeDay": 185.27918781725862,
    "type": "compound",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_flat_Sushi_emission",
    "apr": 8.888,
    "boostable": false,
    "harvestable": false,
    "maxApr": 8.888,
    "minApr": 8.888,
    "name": "Sushi Rewards",
    "oneDay": 8.888,
    "sevenDay": 8.888,
    "thirtyDay": 8.888,
    "threeDay": 8.888,
    "type": "flat_Sushi_emission",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_boosted_bBADGER_emission",
    "apr": 6.969,
    "boostable": false,
    "harvestable": false,
    "maxApr": 6.969,
    "minApr": 6.969,
    "name": "Badger Rewards",
    "oneDay": 6.969,
    "sevenDay": 6.969,
    "thirtyDay": 6.969,
    "threeDay": 6.969,
    "type": "boosted_bBADGER_emission",
  },
]
`;

exports[`vaults.utils getVaultPerformance requests standard vault performance returns value sources from standard methods 1`] = `
Array [
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_precompound",
    "apr": 490142857142857200,
    "boostable": false,
    "harvestable": true,
    "maxApr": 490142857142857200,
    "minApr": 490142857142857200,
    "name": "Vault Compounding",
    "oneDay": 490142857142857200,
    "sevenDay": 490142857142857200,
    "thirtyDay": 490142857142857200,
    "threeDay": 490142857142857200,
    "type": "precompound",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_compound",
    "apr": Infinity,
    "boostable": false,
    "harvestable": false,
    "maxApr": Infinity,
    "minApr": Infinity,
    "name": "Vault Compounding",
    "oneDay": Infinity,
    "sevenDay": Infinity,
    "thirtyDay": Infinity,
    "threeDay": Infinity,
    "type": "compound",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_flat_bCVXCRV_emission",
    "apr": 1242807484412.0166,
    "boostable": false,
    "harvestable": false,
    "maxApr": 1242807484412.0166,
    "minApr": 1242807484412.0166,
    "name": "bCVXCRV Rewards",
    "oneDay": 1242807484412.0166,
    "sevenDay": 1242807484412.0166,
    "thirtyDay": 1242807484412.0166,
    "threeDay": 1242807484412.0166,
    "type": "flat_bCVXCRV_emission",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_bcvxcrv_compounding",
    "apr": 1.522862505950683e+202,
    "boostable": false,
    "harvestable": false,
    "maxApr": 1.522862505950683e+202,
    "minApr": 1.522862505950683e+202,
    "name": "bCVXCRV Compounding",
    "oneDay": 1.522862505950683e+202,
    "sevenDay": 1.522862505950683e+202,
    "thirtyDay": 1.522862505950683e+202,
    "threeDay": 1.522862505950683e+202,
    "type": "bcvxcrv_compounding",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_flat_bveCVX_emission",
    "apr": 5559483778809.337,
    "boostable": false,
    "harvestable": false,
    "maxApr": 5559483778809.337,
    "minApr": 5559483778809.337,
    "name": "bveCVX Rewards",
    "oneDay": 5559483778809.337,
    "sevenDay": 5559483778809.337,
    "thirtyDay": 5559483778809.337,
    "threeDay": 5559483778809.337,
    "type": "flat_bveCVX_emission",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_bvecvx_compounding",
    "apr": 1.398073221267562e+219,
    "boostable": false,
    "harvestable": false,
    "maxApr": 1.398073221267562e+219,
    "minApr": 1.398073221267562e+219,
    "name": "bveCVX Compounding",
    "oneDay": 1.398073221267562e+219,
    "sevenDay": 1.398073221267562e+219,
    "thirtyDay": 1.398073221267562e+219,
    "threeDay": 1.398073221267562e+219,
    "type": "bvecvx_compounding",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_boosted_bBADGER_emission",
    "apr": 6.969,
    "boostable": false,
    "harvestable": false,
    "maxApr": 6.969,
    "minApr": 6.969,
    "name": "Badger Rewards",
    "oneDay": 6.969,
    "sevenDay": 6.969,
    "thirtyDay": 6.969,
    "threeDay": 6.969,
    "type": "boosted_bBADGER_emission",
  },
]
`;

exports[`vaults.utils getVaultPerformance requests standard vault performance skips all emitted tokens with no price 1`] = `
Array [
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_precompound",
    "apr": 490142857142857200,
    "boostable": false,
    "harvestable": true,
    "maxApr": 490142857142857200,
    "minApr": 490142857142857200,
    "name": "Vault Compounding",
    "oneDay": 490142857142857200,
    "sevenDay": 490142857142857200,
    "thirtyDay": 490142857142857200,
    "threeDay": 490142857142857200,
    "type": "precompound",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_compound",
    "apr": Infinity,
    "boostable": false,
    "harvestable": false,
    "maxApr": Infinity,
    "minApr": Infinity,
    "name": "Vault Compounding",
    "oneDay": Infinity,
    "sevenDay": Infinity,
    "thirtyDay": Infinity,
    "threeDay": Infinity,
    "type": "compound",
  },
  CachedValueSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "addressValueSourceType": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28_boosted_bBADGER_emission",
    "apr": 6.969,
    "boostable": false,
    "harvestable": false,
    "maxApr": 6.969,
    "minApr": 6.969,
    "name": "Badger Rewards",
    "oneDay": 6.969,
    "sevenDay": 6.969,
    "thirtyDay": 6.969,
    "threeDay": 6.969,
    "type": "boosted_bBADGER_emission",
  },
]
`;
